### PR TITLE
KAFKA-17949: Introduce GroupState and replace ShareGroupState

### PR DESCRIPTION
**Pull request by AndrewJSchofield**
This PR introduces the unified `GroupState` enum for all group types from KIP-1043. This PR also removes `ShareGroupState` and begins the work to replace `Admin.listShareGroups` with `Admin.listGroups`. That will complete in a future PR.

Integration tests to be enhanced in a future PR.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by omkreddy — APPROVED** [review]
@AndrewJSchofield Thanks for the PR. LGTM.
I will wait for any other reviews before merging the PR.

**Comment by omkreddy** [comment]
needs rebase due to https://github.com/apache/kafka/pull/17712

**Comment by AndrewJSchofield** [comment]
rebased

**Pull request merged**
